dbms/src/Functions/FunctionDateOrDateTimeAddInterval.h

```diff
@@ -286,20 +377,60 @@ class FunctionDateOrDateTimeAddInterval : public IFunction
                     ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
         }
 
-        if (WhichDataType(arguments[0].type).isDate())
+        switch (arguments[0].type->getTypeId())
         {
-            if (std::is_same_v<decltype(Transform::execute(DataTypeDate::FieldType(), 0, std::declval<DateLUTImpl>())), UInt16>)
-                return std::make_shared<DataTypeDate>();
-            else
-                return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, 2, 0));
+            case TypeIndex::Date:
+                return resolveReturnType<DataTypeDate>(arguments);
+            case TypeIndex::DateTime:
+                return resolveReturnType<DataTypeDateTime>(arguments);
+            case TypeIndex::DateTime64:
+                return resolveReturnType<DataTypeDateTime64>(arguments);
+            default:
+            {
+                // TODO (vnemkov): quick and dirty way to check, remove before merging.
```

Comment:
Well, please don't forget it :)

Reply:
removed that piece, thanks for noticing!